File: docs/source/src/rust/engine/expressions.rs
use radiate::prelude::*;

fn my_fitness_fn(geno: Vec<f32>) -> f32 {
    geno.iter().sum()
}

// Dataset + target function for the speciation showcase (block: example).
fn dataset() -> impl Into<DataSet<f32>> {
    let mut inputs = Vec::new();
    let mut answers = Vec::new();

    let mut input = -1.0;
    for _ in -10..10 {
        input += 0.1;
        inputs.push(vec![input]);
        answers.push(vec![compute(input)]);
    }

    (inputs, answers)
}

fn compute(x: f32) -> f32 {
    4.0 * x.powf(3.0) - 3.0 * x.powf(2.0) + x
}

fn main() {
    // --8<-- [start:building]
    // Select a metric by name (reads last_value by default)
    let score = Expr::select("scores.best");

    // A literal constant
    let threshold = Expr::lit(0.01_f32);
    // --8<-- [end:building]

    // --8<-- [start:aggregations]
    let score = Expr::select("scores.best");

    score.clone().last(); // last recorded value (default)
    score.clone().mean(); // running mean
    score.clone().stddev(); // standard deviation
    score.clone().min(); // running minimum
    score.clone().max(); // running maximum
    score.clone().sum(); // running sum
    score.clone().var(); // variance
    score.clone().skew(); // skewness
    score.clone().count(); // number of values seen
    score.clone().slope(); // linear slope over all accumulated values
    score.clone().unique(); // deduplicated collection

    // Rolling window: aggregate over the last N values only
    score.clone().rolling(50).mean();
    score.clone().rolling(100).slope();
    // --8<-- [end:aggregations]

    // --8<-- [start:comparisons]
    let score = Expr::select("scores.best");
    let index = Expr::select("index");

    score.clone().lt(0.01_f32);
    score.clone().lte(0.01_f32);
    score.clone().gt(0.99_f32);
    score.clone().gte(0.99_f32);
    score.clone().eq(0.5_f32);
    score.clone().ne(0.5_f32);

    // Boolean logic
    score.clone().lt(0.01_f32).and(index.clone().gt(50.0_f32)); // and
    score.clone().lt(0.01_f32).or(index.clone().gt(500.0_f32)); // or
    score.clone().lt(0.01_f32).not(); // not

    // Convenience: between (inclusive)
    score.clone().between(0.0_f32, 1.0_f32);
    // --8<-- [end:comparisons]

    // --8<-- [start:arithmetic]
    let a = Expr::select("scores.best");
    let b = Expr::select("score.volatility");

    a.clone().add(b.clone());
    a.clone().sub(b.clone());
    a.clone().mul(Expr::lit(2.0_f32));
    a.clone().div(b.clone());
    a.clone().pow(Expr::lit(2.0_f32));
    a.clone().neg();
    a.clone().abs();
    a.clone().clamp(Expr::lit(0.0_f32), Expr::lit(1.0_f32));
    // --8<-- [end:arithmetic]

    // --8<-- [start:conditional]
    let expr = Expr::when(Expr::select("scores.best").lt(0.01_f32))
        .then(Expr::select("scores.best").mean())
        .otherwise(Expr::lit(1.0_f32));
    // --8<-- [end:conditional]

    // --8<-- [start:schedule]
    let expr = Expr::every(10)
        .then(Expr::select("scores.best").rolling(10).stddev())
        .otherwise(Expr::select("scores.best"));
    // --8<-- [end:schedule]

    // --8<-- [start:querying]
    // Interpret the time metric as Duration
    Expr::select("time").time().mean();

    // Read count.evaluation as a numeric value
    Expr::select("count.evaluation").count();
    // --8<-- [end:querying]

    // --8<-- [start:limit_expr]
    let engine = GeneticEngine::builder()
        .codec(FloatCodec::vector(10, -5.0..5.0))
        .minimizing()
        .fitness_fn(my_fitness_fn)
        .build();

    // Stop when the rolling mean of the best score drops below 0.01
    let stop_expr = Expr::select("scores.best").rolling(50).mean().lt(0.01_f32);

    let result = engine
        .iter()
        .limit((
            Limit::Expr(stop_expr),
            Limit::Generation(5000), // hard ceiling
        ))
        .last()
        .unwrap();
    // --8<-- [end:limit_expr]

    // --8<-- [start:derived_metrics]
    let score_trend = Expr::select("scores.best").rolling(20).slope();
    let score_cv = Expr::select("scores.best")
        .rolling(20)
        .stddev()
        .div(Expr::select("scores.best").rolling(20).mean());

    let engine = GeneticEngine::builder()
        .codec(FloatCodec::vector(10, -5.0..5.0))
        .minimizing()
        .fitness_fn(my_fitness_fn)
        .register_metrics(vec![("score_trend", score_trend), ("score_cv", score_cv)])
        .build();

    let result = engine.iter().limit(5000).last().unwrap();
    let metrics = result.metrics();

    // Access the derived metric values
    if let Some(m) = metrics.get("score_trend") {
        println!("score_trend: {}", m.last_value());
    }
    // --8<-- [end:derived_metrics]

    // --8<-- [start:derived_metrics_limit]
    let engine = GeneticEngine::builder()
        .codec(FloatCodec::vector(10, -5.0..5.0))
        .minimizing()
        .fitness_fn(my_fitness_fn)
        .register_metrics(vec![(
            "score_trend",
            Expr::select("scores.best").rolling(50).slope(),
        )])
        .build();

    let result = engine
        .iter()
        .limit((
            Limit::Expr(Expr::select("score_trend").abs().lt(0.0001_f32)),
            Limit::Generation(5000),
        ))
        .last()
        .unwrap();
    // --8<-- [end:derived_metrics_limit]

    // --8<-- [start:dynamic_rates]
    let dynamic_rate = Expr::select("score.volatility")
        .rolling(20)
        .mean()
        .clamp(0.01_f32, 0.5_f32);

    let engine = GeneticEngine::builder()
        .codec(FloatCodec::vector(10, -5.0..5.0))
        .minimizing()
        .fitness_fn(my_fitness_fn)
        .alter(alters![
            GaussianMutator::new(dynamic_rate),
            BlendCrossover::new(0.5, 0.5),
        ])
        .build();
    // --8<-- [end:dynamic_rates]

    // --8<-- [start:example]
    random_provider::seed(90);

    let store = vec![
        (NodeType::Input, vec![Op::var(0)]),
        (NodeType::Edge, vec![Op::weight()]),
        (NodeType::Vertex, vec![Op::sub(), Op::mul(), Op::linear()]),
        (NodeType::Output, vec![Op::linear()]),
    ];

    let target_species = 4.0;
    let rolling = target_species as usize;

    let spec_count_signal = Expr::select("count.species")
        .rolling(rolling)
        .mean()
        .div(target_species);

    let spec_dist_signal = Expr::select("species.distance")
        .mean()
        .rolling(rolling)
        .mean()
        .div(target_species);

    let spec_thresh_signal = Expr::select("species.threshold").rolling(rolling).mean();
    let spec_evenness_signal = Expr::select("species.evenness").rolling(rolling).mean();

    let distance_signal = spec_count_signal
        .mul(0.9)
        .add(spec_dist_signal.mul(0.4))
        .add(spec_thresh_signal.mul(0.2))
        .add(spec_evenness_signal.mul(0.1))
        .clamp(0.01, 10.0);

    let engine = GeneticEngine::builder()
        .codec(GraphCodec::directed(1, 1, store))
        .raw_batch_fitness_fn(Regression::new(dataset(), Loss::MSE))
        .minimizing()
        .diversity(NeatDistance::new(1.0, 1.0, 3.0))
        .species_threshold(Rate::Expr(distance_signal))
        .alter(alters!(
            GraphCrossover::new(0.5, 0.5),
            OperationMutator::new(0.07, 0.05),
            GraphMutator::new(0.1, 0.1).allow_recurrent(false)
        ))
        .build();
    // --8<-- [end:example]
}

File: docs/source/src/python/engine/expressions.py
import radiate as rd


# Setup (not shown): a trivial fitness so the engine examples below terminate fast and
# deterministically. A small *non-zero* constant keeps the best score below the 0.01
# limits (so they trip quickly) while keeping derived stats like stddev/mean well-defined.
def my_fitness_fn(x):
    return 0.001


# --8<-- [start:building]
import radiate as rd

# Select a metric by name. By default this reads the last recorded value.
score = rd.Expr.select("scores.best")

# A literal constant
threshold = rd.Expr.lit(0.01)

# The current generation index
gen = rd.Expr.generation()
# --8<-- [end:building]

# --8<-- [start:aggregations]
import radiate as rd

score = rd.Expr.select("scores.best")

score.last()  # last recorded value (default)
score.mean()  # running mean of all values seen
score.stddev()  # standard deviation
score.min()  # running minimum
score.max()  # running maximum
score.sum()  # running sum
score.var()  # variance
score.skew()  # skewness
score.count()  # number of values seen
score.slope()  # linear slope over accumulated values
score.unique()  # deduplicated collection

# Rolling window: aggregate over the last N values only
score.rolling(50).mean()  # mean of the last 50 values
score.rolling(50).stddev()  # std dev of the last 50
score.rolling(100).slope()  # slope over a 100-generation window
# --8<-- [end:aggregations]

# --8<-- [start:comparisons]
import radiate as rd

score = rd.Expr.select("scores.best")

# Comparisons — Python operators work directly. Each produces a Bool expression.
below = score < 0.01
at_most = score <= 0.01
above = score > 0.99
at_least = score >= 0.99
equal = score == 0.5
not_equal = score != 0.5

# Boolean logic combines Bool expressions
both = (score < 0.01) & (rd.Expr.select("index") > 50)  # and
either = (score < 0.01) | (rd.Expr.select("time") > 10.0)  # or
negated = ~(score < 0.01)  # not

# A range check is just two comparisons (an inclusive "between")
in_range = (score >= 0.0) & (score <= 1.0)
# --8<-- [end:comparisons]

# --8<-- [start:arithmetic]
import radiate as rd

a = rd.Expr.select("scores.best")
b = rd.Expr.select("score.volatility")

added = a + b
subtracted = a - b
multiplied = a * 2.0
divided = a / b
squared = a**2
negated = -a
abs(a)
a.clamp(0.0, 1.0)
# --8<-- [end:arithmetic]

# --8<-- [start:conditional]
import radiate as rd

# If the best score is below 0.01, use its mean; otherwise use a fallback literal
expr = (
    rd.Expr.when(rd.Expr.select("scores.best") < 0.01)
    .then(rd.Expr.select("scores.best").mean())
    .otherwise(rd.Expr.lit(1.0))
)
# --8<-- [end:conditional]

# --8<-- [start:schedule]
import radiate as rd

# Compute a rolling stddev but only report it every 10 generations;
# otherwise return the last value.
expr = (
    rd.Expr.every(10)
    .then(rd.Expr.select("scores.best").rolling(10).stddev())
    .otherwise(rd.Expr.select("scores.best"))
)
# --8<-- [end:schedule]

# --8<-- [start:querying]
import radiate as rd

# Mean of the time metric, interpreted as a duration
rd.Expr.select("time").time().mean()

# Count of evaluations as a number
rd.Expr.select("count.evaluation").count()
# --8<-- [end:querying]

# --8<-- [start:limit_expr]
import radiate as rd

engine = rd.Engine.float(10, init_range=(-5.0, 5.0)).fitness(my_fitness_fn).minimizing()

# Stop when the best score has been below 0.01 on average over the last 50 generations
stop_expr = rd.Expr.select("scores.best").rolling(50).mean() < 0.01

result = engine.limit(rd.Limit.expr(stop_expr)).run()
# --8<-- [end:limit_expr]

# Rebuild the engine + stop_expr (not shown) so the combined-limit snippet below is runnable.
engine = rd.Engine.float(10, init_range=(-5.0, 5.0)).fitness(my_fitness_fn).minimizing()
stop_expr = rd.Expr.select("scores.best").rolling(50).mean() < 0.01

# --8<-- [start:limit_combined]
result = engine.limit(
    rd.Limit.expr(stop_expr),
    rd.Limit.generations(5000),  # hard ceiling
).run()
# --8<-- [end:limit_combined]

# --8<-- [start:derived_metrics]
import radiate as rd

score_trend = rd.Expr.select("scores.best").rolling(20).slope().debug()
score_cv = (
    rd.Expr.select("scores.best").rolling(20).stddev()
    / rd.Expr.select("scores.best").rolling(20).mean()
)

engine = (
    rd.Engine.float(10, init_range=(-5.0, 5.0))
    .fitness(my_fitness_fn)
    .minimizing()
    .metrics(
        score_trend=score_trend,
        score_cv=score_cv,
    )
    .limit(rd.Limit.generations(500))
)

# These metrics are now available in every generation result
result = engine.run()
metrics = result.metrics()
print(metrics["score_trend"].value_last())
print(metrics["score_cv"].value_last())
# --8<-- [end:derived_metrics]

# --8<-- [start:derived_metrics_limit]
# Register a trend metric, then stop when it flattens
engine = (
    rd.Engine.float(10, init_range=(-5.0, 5.0))
    .fitness(my_fitness_fn)
    .minimizing()
    .metrics(score_trend=rd.Expr.select("scores.best").rolling(50).slope())
    .limit(
        rd.Limit.expr(abs(rd.Expr.select("score_trend")) < 0.0001),
        rd.Limit.generations(5000),
    )
)

result = engine.run()
# --8<-- [end:derived_metrics_limit]

# --8<-- [start:dynamic_rates]
import radiate as rd

# Start aggressive, decay as volatility drops
dynamic_rate = rd.Expr.select("score.volatility").rolling(20).mean().clamp(0.01, 0.5)

engine = (
    rd.Engine.float(10, init_range=(-5.0, 5.0))
    .fitness(my_fitness_fn)
    .minimizing()
    .alters(
        rd.Mutate.gaussian(rate=dynamic_rate),
        rd.Cross.blend(rate=0.5),
    )
)
# --8<-- [end:dynamic_rates]
